keps/sig-cluster-lifecycle/kubeadm/20190424-kubeadm-for-windows.md

```diff
@@ -0,0 +1,287 @@
+
+It should be noted that this PR only proposes enablement of support for Windows worker nodes using kubeadm.
+
+## Motivation
+
+The motivation for this KEP is to provide a tool to allow users to take a Windows machine and join it to an existing Kubernetes cluster with a single command. The user should also be able to reset the node with a single command. We will also provide the tools necessary to install prerequisites and to bring the Windows node to a Ready state.
+
+### Goals
+
+* Create and maintain a Powershell script to install and run Kubernetes prerequisites on Windows
+* Support kubeadm join for Windows
+* Support kubeadm reset for Windows
+* Create and maintain a Powershell script to deploy Windows CNIs
+
+### Non-Goals
+
+* Installing the Container Runtime (e.g. Docker or containerd)
+* Implement kubeadm init for Windows
+* Implement kubeadm join --control-plane
+* Supporting upgrades using kubeadm upgrade (to be revisited for Beta)
+* Running kube-proxy as a DaemonSet on Windows
+* Running Flannel as a DaemonSet on Windows
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+A user will have a Windows machine that they want to join to an existing Kubernetes cluster.
+
+To start, the user will download a set of required binaries (kubelet, kube-proxy, kubeadm, kubectl, Flannel) using a script. The same script will also wrap kubeadm execution.
+
+After executing the script, the kubelet and kube-proxy will be registered as Windows services.
+
+Next, the script will run "kubeadm join ..." to join the node to the cluster. In this step kubeadm will run preflight checks and proceed with the regular join process.
+
+kubeadm will then restart the kubelet service using flags that kubeadm fed to the Windows service. kubeadm will proceed to bootstrap the node. After this process is finished the node should show with the status of NotReady.
+
+The same script will then configure FlannelD. The script will (re-)register FlannelD as a service and start it with the correct configuration, optionally using parameters from kubeadm.
+
+kube-proxy will do the same steps as FlannelD shortly after.
+
+After these steps, the node status should be Ready.
+
+### Implementation Details/Notes/Constraints
+
+#### Kubeadm manages the kubelet start / stop as a service
+
+Kubeadm has a “kubelet” phase that can start/stop the [kubelet service](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/kubelet/kubelet.go).
+
+Kubeadm already supports running the kubelet as a Windows service. This support was added by these two PRs: [60144](https://github.com/kubernetes/kubernetes/pull/60144), [53553](https://github.com/kubernetes/kubernetes/pull/53553).
+
+This existing support will be leveraged by this proposal.
+
```

Comment:
Please add the following in a new paragraph here:

> Currently `kubeadm join` downloads a KubeletConfiguration "component config" from a ConfigMap (that is created during `kubeadm init`) and writes it on all joining nodes. Then each node proceeds to add it's overrides via kubelet flags (that take precedence over config). This process has to be modified so that this KubeletConfiguration is defaulted properly for Windows nodes. One proposed solution is to upload the defaulted KubeletConfiguration for Windows in a separate ConfigMap, but this solution is a subject to change.